pkg/cloud/gcpprovider.go

```diff
@@ -1349,12 +1356,27 @@ func (gcp *GCP) AllNodePricing() (interface{}, error) {
 // NodePricing returns GCP pricing data for a single node
 func (gcp *GCP) NodePricing(key Key) (*Node, error) {
 	gcp.DownloadPricingDataLock.RLock()
-	defer gcp.DownloadPricingDataLock.RUnlock()
 	if n, ok := gcp.Pricing[key.Features()]; ok {
 		klog.V(4).Infof("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
 		n.Node.BaseCPUPrice = gcp.BaseCPUPrice
+		gcp.DownloadPricingDataLock.RUnlock()
 		return n.Node, nil
+	} else if _, ok := gcp.ValidPricingKeys[key.Features()]; ok {
+		gcp.DownloadPricingDataLock.RUnlock()
+		err := gcp.DownloadPricingData()
+		if err != nil {
+			return nil, fmt.Errorf("Download pricing data failed: %s", err.Error())
+		}
+		if n, ok := gcp.Pricing[key.Features()]; ok {
+			klog.V(4).Infof("Returning pricing for node %s: %+v from SKU %s", key, n.Node, n.Name)
+			n.Node.BaseCPUPrice = gcp.BaseCPUPrice
+			return n.Node, nil
+		}
```

Comment:
This code should be lock protected by a read lock, but since you already have acquired a write lock, it'd be optimal to just hold onto it beyond the `DownloadPricingData()`. 

I think maybe what you should here is actually split out the non-lock code from `DownloadPricingData()` into a new method called `downloadPricing()` or something similar. Have `DownloadPricingData()` perform the locking and call `downloadPricing()`. Then above these lines, instead of calling `DownloadPricingData()`, acquire the readwrite lock via `gcp.DownloadPricingLock.Lock()`, call the `downloadPricing()` function, then call `gcp.DownloadPricingLock.Unlock()` before the `return` in each of these conditionals.